File: 3-HARD/Python/10-word-break/proposed-solution/word_break.py
from typing import List, Set, Dict

class WordBreak:
    def __init__(self, s: str, word_dict: Set[str]):
        self.s = s
        self.word_dict = word_dict
        self.memo: Dict[int, List[str]] = {}

    def word_break(self, start: int = 0) -> List[str]:
        if start == len(self.s):
            return [""]

        if start in self.memo:
            return self.memo[start]

        sentences = []
        for end in range(start + 1, len(self.s) + 1):
            word = self.s[start:end]
            if word in self.word_dict:
                for subsentence in self.word_break(end):
                    sentence = word + (" " + subsentence if subsentence else "")
                    sentences.append(sentence)

        self.memo[start] = sentences
        return sentences

File: 3-HARD/JavaScript/10-word-break/proposed-solution/WordBreak.js
// WordBreak.js - Word Break problem using recursion with memoization in JavaScript

class WordBreak {
  constructor(s, wordDict) {
    this.s = s;
    this.wordDict = new Set(wordDict);
    this.memo = new Map();
  }

  wordBreak(start = 0) {
    if (start === this.s.length) {
      return [""];
    }

    if (this.memo.has(start)) {
      return this.memo.get(start);
    }

    const sentences = [];
    for (let end = start + 1; end <= this.s.length; end++) {
      const word = this.s.substring(start, end);
      if (this.wordDict.has(word)) {
        const subsentences = this.wordBreak(end);
        for (const subsentence of subsentences) {
          const sentence = word + (subsentence ? " " + subsentence : "");
          sentences.push(sentence);
        }
      }
    }

    this.memo.set(start, sentences);
    return sentences;
  }
}

export default WordBreak;
